{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a2ad3642",
   "metadata": {},
   "source": [
    "# Beak Alignments Tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "798b70d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/htmek-processing/lib/python3.13/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n",
      "/opt/anaconda3/envs/htmek-processing/lib/python3.13/site-packages/seqlogo/seqlogo.py:6: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import seaborn as sns\n",
    "from Bio import AlignIO\n",
    "import beak.alignments.utils"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab150a99",
   "metadata": {},
   "source": [
    "First, let's load an alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0799b2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = '/Users/micaholivas/Library/CloudStorage/GoogleDrive-mbolivas@stanford.edu/.shortcut-targets-by-id/1-794qIs0qT1G5m39ZBPHOeCKSxeH3Bi8/Designed_AcyPs_manuscript/Data/alignments/AcyP_orthologs/output/aligned_UniProt_29k_plus_SFL.fasta'\n",
    "aln = AlignIO.read(filepath, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a64ceddd",
   "metadata": {},
   "source": [
    "Now we'll \"ungap\" the alignment by removing significantly gapped positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "52cc0efc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<<class 'Bio.Align.MultipleSeqAlignment'> instance (25143 records of length 81) at 110657390>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aln = beak.alignments.utils.ungap_aln(aln)\n",
    "aln"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0047fa60",
   "metadata": {},
   "source": [
    "Great! Let's see what the consensus sequence is from this alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b542e52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--MVRVHAIVSGRVQGVGFRYFTRREARELLTGWVRNLPDGVEVVAEG---AV-ALLEWL-P-AARVD-VEV----FEIR-\n"
     ]
    }
   ],
   "source": [
    "consensus = beak.alignments.utils.get_consensus(aln)\n",
    "print(consensus)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7a99362",
   "metadata": {},
   "source": [
    "Let's make a position-specific scoring matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5075a04c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>R</th>\n",
       "      <th>N</th>\n",
       "      <th>D</th>\n",
       "      <th>C</th>\n",
       "      <th>Q</th>\n",
       "      <th>E</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>...</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>P</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>V</th>\n",
       "      <th>-</th>\n",
       "      <th>cons_i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.034284</td>\n",
       "      <td>0.011057</td>\n",
       "      <td>0.026488</td>\n",
       "      <td>0.030267</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.014557</td>\n",
       "      <td>0.056954</td>\n",
       "      <td>0.022591</td>\n",
       "      <td>0.007437</td>\n",
       "      <td>0.004693</td>\n",
       "      <td>...</td>\n",
       "      <td>0.210357</td>\n",
       "      <td>0.001273</td>\n",
       "      <td>0.025375</td>\n",
       "      <td>0.066698</td>\n",
       "      <td>0.035079</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>0.008074</td>\n",
       "      <td>0.418367</td>\n",
       "      <td>0.346842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.072108</td>\n",
       "      <td>0.030784</td>\n",
       "      <td>0.031500</td>\n",
       "      <td>0.065943</td>\n",
       "      <td>0.000676</td>\n",
       "      <td>0.019926</td>\n",
       "      <td>0.095255</td>\n",
       "      <td>0.043710</td>\n",
       "      <td>0.012767</td>\n",
       "      <td>0.004455</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117568</td>\n",
       "      <td>0.001591</td>\n",
       "      <td>0.034005</td>\n",
       "      <td>0.088136</td>\n",
       "      <td>0.046216</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>0.012449</td>\n",
       "      <td>0.283220</td>\n",
       "      <td>0.211263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.026329</td>\n",
       "      <td>0.075687</td>\n",
       "      <td>0.045460</td>\n",
       "      <td>0.052261</td>\n",
       "      <td>0.002187</td>\n",
       "      <td>0.042636</td>\n",
       "      <td>0.043074</td>\n",
       "      <td>0.006682</td>\n",
       "      <td>0.012926</td>\n",
       "      <td>0.057829</td>\n",
       "      <td>...</td>\n",
       "      <td>0.249374</td>\n",
       "      <td>0.005727</td>\n",
       "      <td>0.024898</td>\n",
       "      <td>0.026687</td>\n",
       "      <td>0.054170</td>\n",
       "      <td>0.001352</td>\n",
       "      <td>0.006960</td>\n",
       "      <td>0.039892</td>\n",
       "      <td>0.048244</td>\n",
       "      <td>0.160818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.060931</td>\n",
       "      <td>0.072147</td>\n",
       "      <td>0.006841</td>\n",
       "      <td>0.005528</td>\n",
       "      <td>0.008034</td>\n",
       "      <td>0.058744</td>\n",
       "      <td>0.065704</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>0.007159</td>\n",
       "      <td>0.128067</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033608</td>\n",
       "      <td>0.008551</td>\n",
       "      <td>0.010420</td>\n",
       "      <td>0.027443</td>\n",
       "      <td>0.050591</td>\n",
       "      <td>0.000955</td>\n",
       "      <td>0.007915</td>\n",
       "      <td>0.217476</td>\n",
       "      <td>0.021915</td>\n",
       "      <td>0.206309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.130812</td>\n",
       "      <td>0.408503</td>\n",
       "      <td>0.009983</td>\n",
       "      <td>0.000239</td>\n",
       "      <td>0.097960</td>\n",
       "      <td>0.028676</td>\n",
       "      <td>0.008511</td>\n",
       "      <td>0.014597</td>\n",
       "      <td>0.035238</td>\n",
       "      <td>0.013284</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>0.001114</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.076522</td>\n",
       "      <td>0.088971</td>\n",
       "      <td>0.001273</td>\n",
       "      <td>0.003102</td>\n",
       "      <td>0.011534</td>\n",
       "      <td>0.019329</td>\n",
       "      <td>0.340924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>0.007636</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>0.001710</td>\n",
       "      <td>0.004216</td>\n",
       "      <td>0.002426</td>\n",
       "      <td>0.001153</td>\n",
       "      <td>0.033807</td>\n",
       "      <td>0.003500</td>\n",
       "      <td>0.021437</td>\n",
       "      <td>0.004216</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004455</td>\n",
       "      <td>0.635087</td>\n",
       "      <td>0.007040</td>\n",
       "      <td>0.004375</td>\n",
       "      <td>0.003062</td>\n",
       "      <td>0.007597</td>\n",
       "      <td>0.089369</td>\n",
       "      <td>0.004972</td>\n",
       "      <td>0.120113</td>\n",
       "      <td>0.542961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>0.031898</td>\n",
       "      <td>0.126238</td>\n",
       "      <td>0.018534</td>\n",
       "      <td>0.050193</td>\n",
       "      <td>0.000994</td>\n",
       "      <td>0.044744</td>\n",
       "      <td>0.202641</td>\n",
       "      <td>0.025136</td>\n",
       "      <td>0.031261</td>\n",
       "      <td>0.015233</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003142</td>\n",
       "      <td>0.006006</td>\n",
       "      <td>0.000517</td>\n",
       "      <td>0.113829</td>\n",
       "      <td>0.050392</td>\n",
       "      <td>0.001153</td>\n",
       "      <td>0.009267</td>\n",
       "      <td>0.031420</td>\n",
       "      <td>0.168794</td>\n",
       "      <td>0.195799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>0.010381</td>\n",
       "      <td>0.013006</td>\n",
       "      <td>0.000239</td>\n",
       "      <td>0.000597</td>\n",
       "      <td>0.002864</td>\n",
       "      <td>0.042676</td>\n",
       "      <td>0.056676</td>\n",
       "      <td>0.000915</td>\n",
       "      <td>0.000517</td>\n",
       "      <td>0.515730</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012687</td>\n",
       "      <td>0.002864</td>\n",
       "      <td>0.001710</td>\n",
       "      <td>0.005210</td>\n",
       "      <td>0.056795</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>0.002386</td>\n",
       "      <td>0.238913</td>\n",
       "      <td>0.014676</td>\n",
       "      <td>0.492962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>0.041363</td>\n",
       "      <td>0.404725</td>\n",
       "      <td>0.008631</td>\n",
       "      <td>0.006403</td>\n",
       "      <td>0.003182</td>\n",
       "      <td>0.019131</td>\n",
       "      <td>0.034960</td>\n",
       "      <td>0.031778</td>\n",
       "      <td>0.018971</td>\n",
       "      <td>0.052659</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007278</td>\n",
       "      <td>0.004335</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.018614</td>\n",
       "      <td>0.053057</td>\n",
       "      <td>0.005091</td>\n",
       "      <td>0.014199</td>\n",
       "      <td>0.070000</td>\n",
       "      <td>0.015710</td>\n",
       "      <td>0.288950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>0.024937</td>\n",
       "      <td>0.120908</td>\n",
       "      <td>0.027085</td>\n",
       "      <td>0.011653</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>0.010898</td>\n",
       "      <td>0.014477</td>\n",
       "      <td>0.049039</td>\n",
       "      <td>0.041920</td>\n",
       "      <td>0.004176</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001989</td>\n",
       "      <td>0.036591</td>\n",
       "      <td>0.062681</td>\n",
       "      <td>0.019131</td>\n",
       "      <td>0.006881</td>\n",
       "      <td>0.014795</td>\n",
       "      <td>0.200215</td>\n",
       "      <td>0.006205</td>\n",
       "      <td>0.296941</td>\n",
       "      <td>0.263529</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A         R         N         D         C         Q         E  \\\n",
       "0   0.034284  0.011057  0.026488  0.030267  0.000437  0.014557  0.056954   \n",
       "1   0.072108  0.030784  0.031500  0.065943  0.000676  0.019926  0.095255   \n",
       "2   0.026329  0.075687  0.045460  0.052261  0.002187  0.042636  0.043074   \n",
       "3   0.060931  0.072147  0.006841  0.005528  0.008034  0.058744  0.065704   \n",
       "4   0.130812  0.408503  0.009983  0.000239  0.097960  0.028676  0.008511   \n",
       "..       ...       ...       ...       ...       ...       ...       ...   \n",
       "76  0.007636  0.003977  0.001710  0.004216  0.002426  0.001153  0.033807   \n",
       "77  0.031898  0.126238  0.018534  0.050193  0.000994  0.044744  0.202641   \n",
       "78  0.010381  0.013006  0.000239  0.000597  0.002864  0.042676  0.056676   \n",
       "79  0.041363  0.404725  0.008631  0.006403  0.003182  0.019131  0.034960   \n",
       "80  0.024937  0.120908  0.027085  0.011653  0.000756  0.010898  0.014477   \n",
       "\n",
       "           G         H         I  ...         M         F         P         S  \\\n",
       "0   0.022591  0.007437  0.004693  ...  0.210357  0.001273  0.025375  0.066698   \n",
       "1   0.043710  0.012767  0.004455  ...  0.117568  0.001591  0.034005  0.088136   \n",
       "2   0.006682  0.012926  0.057829  ...  0.249374  0.005727  0.024898  0.026687   \n",
       "3   0.003977  0.007159  0.128067  ...  0.033608  0.008551  0.010420  0.027443   \n",
       "4   0.014597  0.035238  0.013284  ...  0.001949  0.001114  0.000398  0.076522   \n",
       "..       ...       ...       ...  ...       ...       ...       ...       ...   \n",
       "76  0.003500  0.021437  0.004216  ...  0.004455  0.635087  0.007040  0.004375   \n",
       "77  0.025136  0.031261  0.015233  ...  0.003142  0.006006  0.000517  0.113829   \n",
       "78  0.000915  0.000517  0.515730  ...  0.012687  0.002864  0.001710  0.005210   \n",
       "79  0.031778  0.018971  0.052659  ...  0.007278  0.004335  0.000398  0.018614   \n",
       "80  0.049039  0.041920  0.004176  ...  0.001989  0.036591  0.062681  0.019131   \n",
       "\n",
       "           T         W         Y         V         -    cons_i  \n",
       "0   0.035079  0.000437  0.000318  0.008074  0.418367  0.346842  \n",
       "1   0.046216  0.000080  0.000756  0.012449  0.283220  0.211263  \n",
       "2   0.054170  0.001352  0.006960  0.039892  0.048244  0.160818  \n",
       "3   0.050591  0.000955  0.007915  0.217476  0.021915  0.206309  \n",
       "4   0.088971  0.001273  0.003102  0.011534  0.019329  0.340924  \n",
       "..       ...       ...       ...       ...       ...       ...  \n",
       "76  0.003062  0.007597  0.089369  0.004972  0.120113  0.542961  \n",
       "77  0.050392  0.001153  0.009267  0.031420  0.168794  0.195799  \n",
       "78  0.056795  0.000795  0.002386  0.238913  0.014676  0.492962  \n",
       "79  0.053057  0.005091  0.014199  0.070000  0.015710  0.288950  \n",
       "80  0.006881  0.014795  0.200215  0.006205  0.296941  0.263529  \n",
       "\n",
       "[81 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pssm = beak.alignments.utils.alignment_to_pssm(aln, freq=True)\n",
    "pssm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cacaa4c",
   "metadata": {},
   "source": [
    "Now, for a query sequence, get the frequency of each residue at each aligned position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f4d046",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[7.43745774e-03 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 9.99960227e-01 0.00000000e+00 1.00000000e+00\n",
      "  0.00000000e+00 3.97725013e-05 0.00000000e+00 0.00000000e+00\n",
      "  1.97271606e-02 2.04828382e-02 9.10790280e-03 2.74589349e-01\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 8.27268027e-03 7.56472975e-02\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  3.97725013e-05 5.96587519e-04 2.33862308e-02 7.95450026e-05\n",
      "  9.98289782e-03 1.19317504e-04 3.72270612e-02 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 3.26929961e-02 2.93521060e-02\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 4.60525792e-01 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 4.93179016e-03\n",
      "  4.82559758e-01 3.04657360e-02 6.38348646e-02 2.02442032e-02\n",
      "  4.97156266e-03 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00 0.00000000e+00]]\n"
     ]
    }
   ],
   "source": [
    "my_seq = \"MSTAQSLKSVDYEVFGRVQGVCFRMYTEDEARKIGVVGWVKNTSKGTVTGQVQGPEDKVNSMKSWLSKVGSPSSRIDRTNFSNEKTISKLEYSNFSIRY\"\n",
    "arr = beak.alignments.utils.single_sequence_aln_frequencies(my_seq, pssm, check_positions=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b522873b",
   "metadata": {},
   "source": [
    "Get the conservation of each position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e74c2bba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "conservation = beak.alignments.utils.conservation_from_aln(aln)\n",
    "conservation = beak.alignments.utils.single_sequence_aln_frequencies(my_seq, pssm, check_positions=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bd5214e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize=(8,0.5), dpi=200)\n",
    "# plt.imshow(conservation.reshape(86,1).T, aspect='auto')\n",
    "plt.imshow(conservation.reshape(99,1).T, aspect='auto')\n",
    "plt.colorbar(shrink=1, aspect=2)\n",
    "plt.yticks([])\n",
    "plt.xlabel('Alignment Position')\n",
    "axs.xaxis.tick_top()\n",
    "axs.xaxis.set_label_position('top')\n",
    "plt.tick_params(axis='x', which='major', pad=-0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bac5d65",
   "metadata": {},
   "source": [
    "Use PSSM to get composition of sequence at position i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5572b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "beak.alignments.utils.aln_to_dict(aln)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "680a68c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pssm.iloc[37]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bbc67902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Taxonomic lineage</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P30131</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>HYPF_ECOLI</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>750</td>\n",
       "      <td>cellular organisms (no rank), Bacteria (superk...</td>\n",
       "      <td>MAKNTSCGVQLRIRGKVQGVGFRPFVWQLAQQLNLHGDVCNDGDGV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P75990</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>BLUF_ECOLI</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>403</td>\n",
       "      <td>cellular organisms (no rank), Bacteria (superk...</td>\n",
       "      <td>MLTTLIYRSHIRDDEPVKKIEEMVSIANRRNMQSDVTGILLFNGSH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q8S9F1</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>PCYAB_EUGGR</td>\n",
       "      <td>Euglena gracilis</td>\n",
       "      <td>859</td>\n",
       "      <td>cellular organisms (no rank), Eukaryota (super...</td>\n",
       "      <td>MYILVWKKGQQIKTFHTLDEAAQFKAASNIDEAQMFSVTVAPAISA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q8S9F2</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>PCYAA_EUGGR</td>\n",
       "      <td>Euglena gracilis</td>\n",
       "      <td>1019</td>\n",
       "      <td>cellular organisms (no rank), Eukaryota (super...</td>\n",
       "      <td>MYILVWKEGQQIRTFQDLEECGQFQTASNITDGQIFSINVTPTMSK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P00820</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ACYP2_RABIT</td>\n",
       "      <td>Oryctolagus cuniculus (Rabbit)</td>\n",
       "      <td>99</td>\n",
       "      <td>cellular organisms (no rank), Eukaryota (super...</td>\n",
       "      <td>MSTAGPLKSVDYEVFGRVQGVCFRMYTEGEAKKIGVVGWVKNTSKG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>Q9ZBQ3</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ACYP_STRCO</td>\n",
       "      <td>Streptomyces coelicolor (strain ATCC BAA-471 /...</td>\n",
       "      <td>93</td>\n",
       "      <td>cellular organisms (no rank), Bacteria (superk...</td>\n",
       "      <td>MSEDVRLVAWVRGQVQGVGFRWFTRARALELGGMSGFALNLGDGRV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>Q57973</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Y553_METJA</td>\n",
       "      <td>Methanocaldococcus jannaschii (strain ATCC 430...</td>\n",
       "      <td>153</td>\n",
       "      <td>cellular organisms (no rank), Archaea (superki...</td>\n",
       "      <td>MITTYELIIYGRVQHVGFRDRIEHIGRGLGISGVVYNHKDGTVRIL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>Q58219</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Y809_METJA</td>\n",
       "      <td>Methanocaldococcus jannaschii (strain ATCC 430...</td>\n",
       "      <td>167</td>\n",
       "      <td>cellular organisms (no rank), Archaea (superki...</td>\n",
       "      <td>MVIIMPTTYELKIYGKVQHVGFRDRIENIGKGLGINGIIYNYKDGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>Q58727</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Y1331_METJA</td>\n",
       "      <td>Methanocaldococcus jannaschii (strain ATCC 430...</td>\n",
       "      <td>144</td>\n",
       "      <td>cellular organisms (no rank), Archaea (superki...</td>\n",
       "      <td>MVASMATTYELRIYGNVECAEFIDKVESLGKLLDVNGVVYVYKDSV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>Q58800</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Y1405_METJA</td>\n",
       "      <td>Methanocaldococcus jannaschii (strain ATCC 430...</td>\n",
       "      <td>75</td>\n",
       "      <td>cellular organisms (no rank), Archaea (superki...</td>\n",
       "      <td>MVKFMPTTYEIIIYGRIQHVGFRERIENLGHALGIDGIVYNYEDGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>298 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Entry  Reviewed   Entry Name  \\\n",
       "0    P30131  reviewed   HYPF_ECOLI   \n",
       "1    P75990  reviewed   BLUF_ECOLI   \n",
       "2    Q8S9F1  reviewed  PCYAB_EUGGR   \n",
       "3    Q8S9F2  reviewed  PCYAA_EUGGR   \n",
       "4    P00820  reviewed  ACYP2_RABIT   \n",
       "..      ...       ...          ...   \n",
       "293  Q9ZBQ3  reviewed   ACYP_STRCO   \n",
       "294  Q57973  reviewed   Y553_METJA   \n",
       "295  Q58219  reviewed   Y809_METJA   \n",
       "296  Q58727  reviewed  Y1331_METJA   \n",
       "297  Q58800  reviewed  Y1405_METJA   \n",
       "\n",
       "                                              Organism  Length  \\\n",
       "0                        Escherichia coli (strain K12)     750   \n",
       "1                        Escherichia coli (strain K12)     403   \n",
       "2                                     Euglena gracilis     859   \n",
       "3                                     Euglena gracilis    1019   \n",
       "4                       Oryctolagus cuniculus (Rabbit)      99   \n",
       "..                                                 ...     ...   \n",
       "293  Streptomyces coelicolor (strain ATCC BAA-471 /...      93   \n",
       "294  Methanocaldococcus jannaschii (strain ATCC 430...     153   \n",
       "295  Methanocaldococcus jannaschii (strain ATCC 430...     167   \n",
       "296  Methanocaldococcus jannaschii (strain ATCC 430...     144   \n",
       "297  Methanocaldococcus jannaschii (strain ATCC 430...      75   \n",
       "\n",
       "                                     Taxonomic lineage  \\\n",
       "0    cellular organisms (no rank), Bacteria (superk...   \n",
       "1    cellular organisms (no rank), Bacteria (superk...   \n",
       "2    cellular organisms (no rank), Eukaryota (super...   \n",
       "3    cellular organisms (no rank), Eukaryota (super...   \n",
       "4    cellular organisms (no rank), Eukaryota (super...   \n",
       "..                                                 ...   \n",
       "293  cellular organisms (no rank), Bacteria (superk...   \n",
       "294  cellular organisms (no rank), Archaea (superki...   \n",
       "295  cellular organisms (no rank), Archaea (superki...   \n",
       "296  cellular organisms (no rank), Archaea (superki...   \n",
       "297  cellular organisms (no rank), Archaea (superki...   \n",
       "\n",
       "                                              Sequence  \n",
       "0    MAKNTSCGVQLRIRGKVQGVGFRPFVWQLAQQLNLHGDVCNDGDGV...  \n",
       "1    MLTTLIYRSHIRDDEPVKKIEEMVSIANRRNMQSDVTGILLFNGSH...  \n",
       "2    MYILVWKKGQQIKTFHTLDEAAQFKAASNIDEAQMFSVTVAPAISA...  \n",
       "3    MYILVWKEGQQIRTFQDLEECGQFQTASNITDGQIFSINVTPTMSK...  \n",
       "4    MSTAGPLKSVDYEVFGRVQGVCFRMYTEGEAKKIGVVGWVKNTSKG...  \n",
       "..                                                 ...  \n",
       "293  MSEDVRLVAWVRGQVQGVGFRWFTRARALELGGMSGFALNLGDGRV...  \n",
       "294  MITTYELIIYGRVQHVGFRDRIEHIGRGLGISGVVYNHKDGTVRIL...  \n",
       "295  MVIIMPTTYELKIYGKVQHVGFRDRIENIGKGLGINGIIYNYKDGT...  \n",
       "296  MVASMATTYELRIYGNVECAEFIDKVESLGKLLDVNGVVYVYKDSV...  \n",
       "297  MVKFMPTTYEIIIYGRIQHVGFRERIENLGHALGIDGIVYNYEDGT...  \n",
       "\n",
       "[298 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "new_aln_file = '/Users/micaholivas/Downloads/quick_acyps_aligned.fasta'\n",
    "new_aln = AlignIO.read(new_aln_file, \"fasta\")\n",
    "\n",
    "tsv_file = '/Users/micaholivas/Downloads/uniprotkb_acylphosphatase_AND_reviewed_2025_05_02.tsv'\n",
    "df = pd.read_csv(tsv_file, sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d65db7",
   "metadata": {},
   "outputs": [],
   "source": [
    "length_cutoff = 100\n",
    "\n",
    "# Drop sequences longer than 110 AA from df_expanded\n",
    "df = df[df['Length'] <= length_cutoff]\n",
    "\n",
    "# Drop sequences longer than 110 AA (not counting gaps) from new_aln\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "filtered_records = [\n",
    "    record for record in new_aln\n",
    "    if len(str(record.seq).replace(\"-\", \"\")) <= length_cutoff\n",
    "]\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "new_aln = MultipleSeqAlignment(filtered_records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e2b8fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import re\n",
    "\n",
    "def parse_taxonomic_lineage(lineage_str):\n",
    "    # Split by comma, then extract name and rank using regex\n",
    "    items = [item.strip() for item in lineage_str.split(',')]\n",
    "    parsed = {}\n",
    "    for item in items:\n",
    "        match = re.match(r\"(.+?) \\((.+?)\\)$\", item)\n",
    "        if match:\n",
    "            name, rank = match.groups()\n",
    "            parsed[rank] = name\n",
    "        else:\n",
    "            # If no rank, use as is\n",
    "            parsed['no rank'] = item\n",
    "    return parsed\n",
    "\n",
    "# Apply to the column and create a DataFrame\n",
    "tax_df = df['Taxonomic lineage'].apply(parse_taxonomic_lineage).apply(pd.Series)\n",
    "\n",
    "# Concatenate with original DataFrame if needed\n",
    "df_expanded = pd.concat([df, tax_df], axis=1)\n",
    "\n",
    "df_expanded = df_expanded.drop(columns=['Organism', 'Taxonomic lineage'])\n",
    "\n",
    "df_expanded.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2f4659f",
   "metadata": {},
   "source": [
    "Ungap the alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f05e39b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_aln = beak.alignments.utils.ungap_aln(new_aln)\n",
    "for record in new_aln:\n",
    "    print(record.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c138e9e",
   "metadata": {},
   "source": [
    "Now, merge the aligned sequences into the df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31ee2bfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def extract_entry_id(header):\n",
    "    parts = header.split('|')\n",
    "    if len(parts) >= 3:\n",
    "        return parts[1]\n",
    "    else:\n",
    "        return header  # fallback if not in expected format\n",
    "\n",
    "# Step 1: Extract IDs and sequences from new_aln\n",
    "aln_records = [(extract_entry_id(record.id), str(record.seq)) for record in new_aln]\n",
    "aln_df = pd.DataFrame(aln_records, columns=['Entry', 'Aligned_sequence'])\n",
    "\n",
    "# Step 2: Merge with expanded TSV DataFrame\n",
    "merged_df = df_expanded.merge(aln_df, on='Entry', how='left')\n",
    "\n",
    "merged_df.sample(5)\n",
    "# ...existing code..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7ec4896",
   "metadata": {},
   "source": [
    "Now that we've merged our taxonomic information into the df, compute a PSSM for each"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ae62404",
   "metadata": {},
   "outputs": [],
   "source": [
    "pssms_by_tax_rank = beak.alignments.utils.pssms_by_taxon(merged_df, 'kingdom')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73fc6066",
   "metadata": {},
   "outputs": [],
   "source": [
    "import holoviews as hv\n",
    "import panel as pn\n",
    "import pandas as pd\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "hv.extension('bokeh')\n",
    "\n",
    "def interactive_pssm_heatmap(\n",
    "    pssms_by_tax_rank,\n",
    "    consensus,\n",
    "    rank='kingdom',\n",
    "    phosphate_loop={9,11,12,13,14},\n",
    "    catalytic={17,35}\n",
    "):\n",
    "    \"\"\"\n",
    "    Interactive PSSM heatmap explorer by taxonomic rank.\n",
    "\n",
    "    Args:\n",
    "        pssms_by_tax_rank: dict of {taxon: PSSM DataFrame}\n",
    "        consensus: consensus sequence string\n",
    "        rank: taxonomic rank label for y-axis\n",
    "        phosphate_loop: set of positions to highlight as phosphate loop\n",
    "        catalytic: set of positions to highlight as catalytic residues\n",
    "\n",
    "    Returns:\n",
    "        pn.Column Panel layout for interactive exploration\n",
    "    \"\"\"\n",
    "    # Prepare amino acid set\n",
    "    aas = set()\n",
    "    for pssm in pssms_by_tax_rank.values():\n",
    "        aas.update(pssm.columns)\n",
    "    aas = sorted(aas)\n",
    "\n",
    "    # Custom colormap: 0 is white, 1 is blue\n",
    "    cmap = LinearSegmentedColormap.from_list(\"white_blue\", [\"white\", \"blue\"])\n",
    "\n",
    "    def highlight_consensus(consensus, pos):\n",
    "        html = \"\"\n",
    "        for i, aa in enumerate(consensus):\n",
    "            style = \"font-size:18px\"\n",
    "            if i == pos:\n",
    "                style += \";background-color:yellow; color:black; font-weight:bold\"\n",
    "            if i in phosphate_loop:\n",
    "                style += \";color:green; font-weight:bold\"\n",
    "            if i in catalytic:\n",
    "                style += \";color:red; font-weight:bold\"\n",
    "            html += f\"<span style='{style}'>{aa}</span>\"\n",
    "        return f\"<div style='font-family:monospace; word-break:break-all'>{html}</div>\"\n",
    "\n",
    "    def plot_heatmap(position=0):\n",
    "        # Build DataFrame for the selected position\n",
    "        heatmap_data = []\n",
    "        index = []\n",
    "        for sk, pssm in pssms_by_tax_rank.items():\n",
    "            row = []\n",
    "            for aa in aas:\n",
    "                row.append(pssm.iloc[position][aa] if aa in pssm.columns else 0)\n",
    "            heatmap_data.append(row)\n",
    "            index.append(sk)\n",
    "        heatmap_df = pd.DataFrame(heatmap_data, columns=aas, index=index)\n",
    "        tidy = heatmap_df.reset_index().melt(id_vars='index', var_name='AA', value_name='Frequency')\n",
    "        heatmap = hv.HeatMap(tidy, kdims=['AA', 'index'], vdims='Frequency').opts(\n",
    "            cmap=cmap,\n",
    "            colorbar=True,\n",
    "            clim=(0, 1),\n",
    "            xrotation=0,\n",
    "            yrotation=0,\n",
    "            xlabel='Amino Acid',\n",
    "            ylabel=rank.capitalize(),\n",
    "            colorbar_opts={'title': 'Frequency'},\n",
    "            tools=['hover'],\n",
    "            width=800,\n",
    "            height=300,\n",
    "            line_color='black',\n",
    "            show_grid=True,\n",
    "            toolbar='above',\n",
    "            labelled=['x', 'y', 'colorbar'],\n",
    "            xaxis='top',\n",
    "            fontsize={'xticks': 14, 'yticks': 14, 'ylabel': 14, 'xlabel': 14, 'title': 16}\n",
    "        )\n",
    "        return heatmap.opts(title=f\"Aligned Position {position+1}\")\n",
    "\n",
    "    slider = pn.widgets.IntSlider(name='Aligned Position', start=0, end=len(consensus)-1, value=0)\n",
    "\n",
    "    @pn.depends(slider)\n",
    "    def consensus_view(position):\n",
    "        return pn.pane.HTML(highlight_consensus(consensus, position), width=800)\n",
    "\n",
    "    dmap = hv.DynamicMap(pn.bind(plot_heatmap, position=slider))\n",
    "\n",
    "    return pn.Column(\n",
    "        slider,\n",
    "        dmap,\n",
    "        pn.pane.Markdown(\"## Consensus sequence (highlighted position):\"),\n",
    "        consensus_view\n",
    "    )\n",
    "\n",
    "# Example usage:\n",
    "panel = interactive_pssm_heatmap(pssms_by_tax_rank, consensus, rank='kingdom')\n",
    "panel.servable()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8c92588",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df['superkingdom'].unique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "htmek-processing",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
